cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/kafka/KafkaLogProcessorPipeline.java

```diff
@@ -288,147 +289,82 @@ private boolean processMessages(String topic, int partition,
-   */
-  private <T extends Map<Integer, Future<Iterable<MessageAndOffset>>>> T fetchAll(Int2LongMap offsets,
-                                                                                  T fetchFutures) {
-    for (final int partition : config.getPartitions()) {
-      final long offset = offsets.get(partition);
-
-      fetchFutures.put(partition, fetchExecutor.submit(() -> fetchMessages(partition, offset)));
-    }
-
-    return fetchFutures;
+    return process(topic, partition, messages);
   }
 
-  /**
-   * Appends buffered events to appender. If the {@code force} parameter is {@code false}, buffered events
-   * that are older than the buffer milliseconds will be appended and removed from the buffer.
-   * If {@code force} is {@code true}, then at least {@code maxQueueSize * MIN_FREE_FACTOR} events will be appended
-   * and removed, regardless of the event time.
-   *
-   * @return number of events appended to the appender
-   */
-  private int appendEvents(long currentTimeMillis, boolean forced) {
-    long minEventTime = currentTimeMillis - config.getEventDelayMillis();
-    long maxRetainSize = forced ? (long) (config.getMaxBufferSize() * MIN_FREE_FACTOR) : Long.MAX_VALUE;
-
-    TimeEventQueue.EventIterator<ILoggingEvent, OffsetTime> iterator = eventQueue.iterator();
-
-    int eventsAppended = 0;
-    long minDelay = Long.MAX_VALUE;
-    long maxDelay = -1;
+  private boolean process(String topic, int partition, Iterable<MessageAndOffset> messages) {
+    final byte[] processed = {0};
 
-    while (iterator.hasNext()) {
-      ILoggingEvent event = iterator.next();
+    // process all the messages
+    ProcessedEventMetadata<KafkaOffset> metadata = eventQueueProcessor.process(partition,
+                                                                               messages.iterator(), message -> {
+        ProcessorEvent<KafkaOffset> processorEvent = null;
+        try {
+          ILoggingEvent loggingEvent = serializer.fromBytes(message.message().payload());
+          processorEvent = new ProcessorEvent<>(loggingEvent, message.message().payloadSize(),
+                                                new KafkaOffset(message.nextOffset(), loggingEvent.getTimeStamp()));
+        } catch (IOException e) {
+          // This shouldn't happen. In case it happens (e.g. someone published some garbage), just skip the message.
+          LOG.trace("Fail to decode logging event from {}:{} at offset {}. Skipping it.",
+                    topic, partition, message.offset(), e);
+        }
 
-      // If not forced to reduce the event queue size and the current event timestamp is still within the
-      // buffering time, no need to iterate anymore
-      if (eventQueue.getEventSize() <= maxRetainSize && event.getTimeStamp() >= minEventTime) {
-        break;
-      }
+        processed[0] = 1;
+        metricsContext.increment("kafka.bytes.read", message.message().payloadSize());
+        offsets.put(partition, message.nextOffset());
```

Comment:
setting state in a function meant for translating one object into the next isn't a great pattern. The caller here is assuming knowledge about how the process() method is implemented. It's better if this information is returned by the process() method.

Reply:
Now changing it in the iterator when next event is consumed.